File: backend/src/modules/runtime/runtime.service.js
'use strict';

/**
 * LC-ARCH-004 Phase 2: this module is now a thin HTTP CLIENT to the standalone Legion Runtime
 * process (see `runtime/`) — it no longer owns the poll loop, in-memory controller store, or any
 * BACnet/SIM execution itself. Every exported function keeps its original name/signature (same
 * shapes `runtime.controller.js` and the rest of the Server already call) so this remains a
 * boring, compatible change from the API's point of view; only the ownership moved.
 *
 * A Runtime outage never crashes the Server: `runtimeClient` turns connection failures into
 * `HttpError(503)`, which `asyncHandler`/`errorHandler` already turn into a clean JSON error
 * response instead of an uncaught exception.
 */

const { runtimeGet, runtimePost, runtimeProbe } = require('../../lib/runtimeClient');

/** Legacy export retained for any caller still importing it directly (no SIM catalog dependency here anymore). */
const FCU_CONTROLLER_CODE = 'FCU-1';

async function listControllers() {
  return (await runtimeGet('/runtime/controllers')) || [];
}

async function getController(code) {
  return runtimeGet(`/runtime/controllers/${encodeURIComponent(code)}`, { notFoundValue: null });
}

async function listFieldPointsForController(code) {
  const body = await runtimeGet(`/runtime/controllers/${encodeURIComponent(code)}/field-points`, { notFoundValue: null });
  return body ? body.points : null;
}

async function setOnline(code, online) {
  const path = `/runtime/controllers/${encodeURIComponent(code)}/${online ? 'online' : 'offline'}`;
  return runtimePost(path, {}, { notFoundValue: null });
}

async function setSimEnabled(code, enabled) {
  const path = `/runtime/controllers/${encodeURIComponent(code)}/${enabled ? 'start' : 'stop'}`;
  return runtimePost(path, {}, { notFoundValue: null });
}

async function pollNow(code) {
  return runtimePost(`/runtime/controllers/${encodeURIComponent(code)}/poll-now`, {}, { notFoundValue: null });
}

/**
 * WRITE: the only path a live field write reaches a real device through — Server (here) -> Runtime
 * internal API -> BacnetDriver -> device. The backend process never talks to BACnet directly for a
 * normal live write (Engineering commissioning tooling under `/api/runtime/bacnet/*` is separate —
 * see `bacnet.controller.js` — and intentionally still calls the BACnet client directly for
 * pre-deployment testing of arbitrary/undeployed addresses).
 */
async function writePoint(code, fieldPointKey, value, options = {}) {
  return runtimePost(
    `/runtime/controllers/${encodeURIComponent(code)}/write`,
    { fieldPointKey, value, priority: options.priority },
    { notFoundValue: null }
  );
}

async function listDiscoveryDevices(siteId) {
  const qs = siteId ? `?siteId=${encodeURIComponent(siteId)}` : '';
  const body = await runtimeGet(`/runtime/discovery-devices${qs}`);
  return body ? body.devices : [];
}

/**
 * RELOAD: called after `deployRelease`/`rollbackToVersion` (or the LS-100 activation pipeline)
 * activates a new release — tells Runtime to re-resolve its controller store against the freshly
 * materialized `LiveControllerBinding` rows. Never called from an Engineering-side edit.
 *
 * Deliberately does not throw on failure: per LC-ARCH-004, a Runtime reload failure must not roll
 * back the database activation that already committed. The caller logs/reports degraded state and
 * a later successful reload (manual retry, or Runtime's own restart-time reload) catches up.
 */
async function resyncLiveSimBindings() {
  try {
    return await runtimePost('/runtime/reload', {});
  } catch (e) {
    // eslint-disable-next-line no-console
    console.warn('[runtime-client] reload failed — Runtime is unreachable or degraded:', e?.message || e);
    return { ok: false, reachable: false };
  }
}

/** Non-throwing status probe for Server health endpoints and the deploy-path warning above. */
async function getRuntimeHealth() {
  const probe = await runtimeProbe('/health');
  return probe;
}

module.exports = {
  listControllers,
  getController,
  setOnline,
  setSimEnabled,
  pollNow,
  writePoint,
  listDiscoveryDevices,
  listFieldPointsForController,
  resyncLiveSimBindings,
  getRuntimeHealth,
  FCU_CONTROLLER_CODE,
};

File: server/src/main/java/com/legioncontrols/server/runtime/RuntimeService.java
package com.legioncontrols.server.runtime;

import tools.jackson.databind.JsonNode;
import tools.jackson.databind.node.ObjectNode;
import tools.jackson.databind.node.JsonNodeFactory;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.Map;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

/**
 * Java equivalent of backend/src/modules/runtime/runtime.service.ts — a thin HTTP CLIENT to the
 * standalone Legion Runtime process. Every method keeps the same shape/behavior as its Node
 * counterpart (LC-ARCH-004 Phase 2): Runtime owns polling/BACnet/SIM entirely, this Server never
 * regains a direct/in-process implementation.
 */
@Service
public class RuntimeService {

    private static final Logger log = LoggerFactory.getLogger(RuntimeService.class);

    private final RuntimeClient client;

    public RuntimeService(RuntimeClient client) {
        this.client = client;
    }

    private static String encode(String value) {
        return URLEncoder.encode(value, StandardCharsets.UTF_8);
    }

    public JsonNode listControllers() {
        JsonNode result = client.get("/runtime/controllers");
        return result != null ? result : JsonNodeFactory.instance.arrayNode();
    }

    public JsonNode getController(String code) {
        return client.get("/runtime/controllers/" + encode(code), null);
    }

    public JsonNode listFieldPointsForController(String code) {
        JsonNode body = client.get("/runtime/controllers/" + encode(code) + "/field-points", null);
        return body != null ? body.get("points") : null;
    }

    public JsonNode setOnline(String code, boolean online) {
        String path = "/runtime/controllers/" + encode(code) + "/" + (online ? "online" : "offline");
        return client.post(path, Map.of(), null);
    }

    public JsonNode setSimEnabled(String code, boolean enabled) {
        String path = "/runtime/controllers/" + encode(code) + "/" + (enabled ? "start" : "stop");
        return client.post(path, Map.of(), null);
    }

    public JsonNode pollNow(String code) {
        return client.post("/runtime/controllers/" + encode(code) + "/poll-now", Map.of(), null);
    }

    /**
     * WRITE: the only path a live field write reaches a real device through — this Server -> Runtime
     * internal API -> BacnetDriver -> device. Mirrors runtime.service.ts#writePoint exactly.
     */
    public JsonNode writePoint(String code, String fieldPointKey, Object value, Integer priority) {
        ObjectNode payload = JsonNodeFactory.instance.objectNode();
        payload.put("fieldPointKey", fieldPointKey);
        payload.putPOJO("value", value);
        if (priority != null) {
            payload.put("priority", priority);
        }
        return client.post("/runtime/controllers/" + encode(code) + "/write", payload, null);
    }

    public JsonNode listDiscoveryDevices(String siteId) {
        String qs = (siteId != null && !siteId.isBlank()) ? "?siteId=" + encode(siteId) : "";
        JsonNode body = client.get("/runtime/discovery-devices" + qs);
        JsonNode devices = body != null ? body.get("devices") : null;
        return devices != null ? devices : JsonNodeFactory.instance.arrayNode();
    }

    /**
     * RELOAD: tells Runtime to re-resolve its controller store against freshly materialized
     * LiveControllerBinding rows. Deliberately never throws — a reload failure must never roll back
     * the DB activation that already committed (see releases.DeploymentService).
     */
    public JsonNode resyncLiveSimBindings() {
        try {
            return client.post("/runtime/reload", Map.of());
        } catch (Exception e) {
            log.warn("[runtime-client] reload failed — Runtime is unreachable or degraded: {}", e.getMessage());
            ObjectNode degraded = JsonNodeFactory.instance.objectNode();
            degraded.put("ok", false);
            degraded.put("reachable", false);
            return degraded;
        }
    }

    /** Non-throwing status probe for Server health endpoints. */
    public RuntimeClient.RuntimeProbe getRuntimeHealth() {
        return client.probe("/health");
    }
}
